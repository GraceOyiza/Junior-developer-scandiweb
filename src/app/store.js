import { configureStore } from '@reduxjs/toolkit';
import cartReducer from '../feature/cartSlice';
import currencyReducer from '../feature/currencySlice';
import productReducer from '../feature/productSlice';
import categoryReducer from '../feature/categorySlice';

export default configureStore({
  reducer: {
    cart: cartReducer,
    currency: currencyReducer,
    product: productReducer,
    category: categoryReducer,
  },
});