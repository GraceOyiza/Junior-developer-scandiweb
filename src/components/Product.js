import React from 'react';
import { Button } from '../styled/Button.styled';
import { Flex, Image, Name, Brand, Price, Pdpr, Div, Div2} from '../styled/ProductStyled';

import { setProduct, setImageIndex, setVariant } from '../feature/productSlice';
import { addToCart, togglePopUpCart } from '../feature/cartSlice';
import { connect } from 'react-redux';







class Product extends React.Component {
  componentDidMount() {
    const urlPathName = window.location.pathname.split('/');
    const productId = urlPathName[urlPathName.length - 1];
    getProductQuery(productId).then(({ data: { product } }) => {
      this.props.setProduct(product);
      console.log('Product: ', product);
    });
  }

  handleThumbNailClick = e => {
    this.props.setImageIndex(e.target.dataset.index);
  };

  addToCart = () => {
    this.props.addToCart(this.props.product);
    this.props.togglePopUpCart();
  };

  selectVariant = (attribute, selectedValue) => {
    console.log('clicked', attribute, selectedValue);
    this.props.setVariant({ id: attribute, value: selectedValue });
  }

  render() {
    return (
      <Flex>
        <Flex>
          <div>
            {this.props.product?.gallery?.map((imgSrc, index) => (
              <div onClick={this.handleThumbNailClick} data-index={index}>
                <ThumbNail index={index} src={imgSrc} />
              </div>
            ))}
          </div>

          <Div2>
            <Image
              src={
                this.props.product?.gallery &&
                this.props.product.gallery[this.props.imageIndex]
              }
            />
          </Div2>
        </Flex>
        <div>
          <div>
            <Flex>
              <div>
                <Name> {this.props.product.name}</Name>
                <Brand> {this.props.product.brand}</Brand>
                {this.props.product?.attributes?.map((attribute, index) => (
                  <div key={index}>
                    <VariantList
                      title={attribute.name.toUpperCase()}
                      type={attribute.type}
                      variants={attribute.items}
                      selectVariant={this.selectVariant}
                    />
                  </div>
                ))}
                <Price>PRICE:</Price>
                <Pdpr>
                  {this.props.currentCurrency &&
                  this.props.product?.prices?.length
                    ? setPriceBasedOnCurrency(
                        this.props.product.prices,
                        this.props.currentCurrency,
                      )
                    : ''}
                </Pdpr>
                <Button onClick={this.addToCart}>ADD TO CART</Button>
                <Div
                  dangerouslySetInnerHTML={{
                    __html: this.props.product.description,
                  }}
                />
              </div>
            </Flex>
          </div>
        </div>
      </Flex>
    );
  }
}

const mapStateToProps = state => ({
  product: state.product.product,
  imageIndex: state.product.imageIndex,
  currentCurrency: state.currency.currentCurrency,
});
const mapDispatchToProps = {
  setProduct,
  setImageIndex,
  addToCart,
  togglePopUpCart,
  setVariant,
};


export default connect(mapStateToProps, mapDispatchToProps)(Product);






