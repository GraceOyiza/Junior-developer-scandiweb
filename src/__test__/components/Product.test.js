import React from 'react';
import Enzyme, { render } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Provider } from 'react-redux';
import store from '../../app/store';
import { BrowserRouter } from 'react-router-dom';
import Product from '../../components/Product';

Enzyme.configure({ adapter: new Adapter() });

let wrapper;

beforeEach(() => {
  wrapper = render(
    <BrowserRouter>
      <Provider store={store}>
        <Product />
      </Provider>
    </BrowserRouter>,
  );
});

describe('', () => {
  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
