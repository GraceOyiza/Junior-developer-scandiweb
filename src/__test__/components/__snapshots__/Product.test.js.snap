// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` renders correctly 1`] = `
LoadedCheerio {
  "0": <div
    class="sc-gswNZR eZorJe"
  >
    <div
      class="sc-gswNZR eZorJe"
    >
      <div />
      <div
        class="sc-dkrFOg heZvrG"
      >
        <img
          class="sc-hLBbgP iquyTc"
        />
      </div>
    </div>
    <div>
      <div>
        <div
          class="sc-gswNZR eZorJe"
        >
          <div>
            <div
              class="sc-jSUZER eMeJiA"
            >
               
            </div>
            <div
              class="sc-gKPRtg oMuXV"
            >
               
            </div>
            <div
              class="sc-iBYQkv bSBXIm"
            >
              PRICE:
            </div>
            <div
              class="sc-ftTHYK cKdbmk"
            />
            <button
              class="sc-bcXHqe iXndgr"
            >
              ADD TO CART
            </button>
            <div
              class="sc-eDvSVe ikvxKf"
            />
          </div>
        </div>
      </div>
    </div>
  </div>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
